src/utilities/xml/XMLValidator.cpp

```diff
@@ -0,0 +1,438 @@
+
+  return result;
+}
+
+bool XMLValidator::isValid() const {
+  return errors().empty();
+}
+
+void XMLValidator::reset() {
+
+  m_logSink.setThreadId(std::this_thread::get_id());
+
+  m_logSink.resetStringStream();
+
+  m_xmlPath = boost::none;
+
+  m_fullValidationReport.clear();
+}
+
+bool XMLValidator::validate(const openstudio::path& xmlPath) {
+
+  reset();
+
+  auto t_xmlPath = openstudio::filesystem::system_complete(xmlPath);
+
+  if (!openstudio::filesystem::exists(t_xmlPath)) {
+    LOG_AND_THROW("XML File '" << toString(t_xmlPath) << "' does not exist");
+  } else if (!openstudio::filesystem::is_regular_file(t_xmlPath)) {
+    LOG_AND_THROW("XML File '" << toString(t_xmlPath) << "' cannot be opened");
+  }
+  m_xmlPath = t_xmlPath;
+
+  if (m_validatorType == XMLValidatorType::XSD) {
+    return xsdValidate();
+  } else if ((m_validatorType == XMLValidatorType::XSLTSchematron) || (m_validatorType == XMLValidatorType::Schematron)) {
+    return xsltValidate();
+  }
+
+  return false;
+}
+
+bool XMLValidator::xsdValidate() const {
+
+  // schema path
+  auto schema_filename_str = toString(m_schemaPath);
+  const auto* schema_filename = schema_filename_str.c_str();
+
+  // xml path
+  auto xml_filename_str = toString(m_xmlPath.get());
+  const auto* xml_filename = xml_filename_str.c_str();
+
+  // schema parser ptr
+  xmlSchemaParserCtxt* parser_ctxt = xmlSchemaNewParserCtxt(schema_filename);
+
+  // set parser errors
+  detail::ErrorCollector schemaParserErrorCollector;
+  xmlSchemaSetParserErrors(parser_ctxt, detail::callback_messages_error, detail::callback_messages_warning, &schemaParserErrorCollector);
+
+  // schema parser
+  xmlSchema* schema = xmlSchemaParse(parser_ctxt);
+  xmlSchemaValidCtxt* ctxt = xmlSchemaNewValidCtxt(schema);
+
+  // set valid errors
+  detail::ErrorCollector schemaValidErrorCollector;
+  xmlSchemaSetValidErrors(ctxt, detail::callback_messages_error, detail::callback_messages_warning, &schemaValidErrorCollector);
+
+  detail::ErrorCollector parseFileErrorCollector;
+  xmlSetStructuredErrorFunc(&parseFileErrorCollector, detail::callback_messages_structured_error);
+  xmlSetGenericErrorFunc(&parseFileErrorCollector, detail::callback_messages_error);
+
+  // xml doc ptr
+  xmlDoc* doc = xmlParseFile(xml_filename);
+
+  // validate doc
+  int ret = xmlSchemaValidateDoc(ctxt, doc);
+  bool result = false;
+  if (ret > 0) {
+    LOG(Fatal, "Valid instance " << toString(m_xmlPath.get()) << " failed to validate against " << toString(m_schemaPath));
+    result = false;
+  } else if (ret < 0) {
+    LOG(Fatal, "Valid instance " << toString(m_xmlPath.get()) << " got internal error validating against " << toString(m_schemaPath));
+    result = true;
+  } else {
+    result = true;
+  }
+
+  for (auto& logMessage : schemaValidErrorCollector.logMessages) {
+    LOG(logMessage.logLevel(), "xsdValidate.schemaValidError: " + logMessage.logMessage())
+  }
```

Comment:
Here I'm reissuing the logMessages, but really I don't think we really need a Logger and LogSink here, we could just work with a std::vector<LogMessage> member